# **The Next Hot Place to Live in Brooklyn** 

I will try to figure out the next tendy neighborhood in brooklyn by comparing them to them to current trendy areas such as Williamsburg and Park Slope. Then I will find the area that compares to them and has the lowest rent. 

In [1]:
import bs4 as bs

import urllib.request  

import numpy as np 
import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json 

from geopy.geocoders import Nominatim 

import requests 
from pandas.io.json import json_normalize 
import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans
import folium 

print('Libraries imported.')

Libraries imported.


In [2]:
brooklyn_rent_file='brooklyn.csv'
rent_bk=pd.read_csv(brooklyn_rent_file)
rent_bk

Neighborhood  Average Rent
0                             Flatlands          1492
1                           Marine Park          1659
2                               Madison          1659
3                       Gerritsen Beach          1659
4                             Homecrest          1717
5                           Bensonhurst          1795
6                            Bath Beach          1795
7                               Midwood          1825
8                         Fiske Terrace          1825
9                            Kensington          1839
10                            Gravesend          1840
11                        Crown Heights          1853
12                         Borough Park          1898
13                          Ditmas Park          1944
14                  Prospect Park South          1997
15          Prospect - Lefferts Gardens          2009
16                             Flatbush          2009
17                              Wingate          2114
18                       Sheepshead Bay          2131
19                        East Flatbush          2178
20                             Canarsie          2178
21                      Manhattan Beach          2214
22                       Brighton Beach          2214
23                      Windsor Terrace          2361
24                         Clinton Hill          2671
25                   Bedford-Stuyvesant          2784
26                        Prospect Park          2876
27                             Red Hook          2916
28                      Carroll Gardens          2916
29                             Bushwick          3065
30                   Brooklyn Navy Yard          3136
31                              Gowanus          3247
32                         Williamsburg          3247
33                     Prospect Heights          3373
34  Columbia Street Waterfront District          3401
35                          Cobble Hill          3401
36                          Fort Greene          3403
37                         Vinegar Hill          3456
38                                DUMBO          3456
39                    Downtown Brooklyn          3456
40                     Brooklyn Heights          3456
41                          Boerum Hill          3462
42                           Park Slope          3513
43                    Greenwood Heights          3538

check columns

In [3]:
for col in rent_bk.columns: 
    print(col) 

Neighborhood
Average Rent


Get brooklyn data

In [4]:
with open('nyu-2451-34572-geojson.json') as json_data:
    new_york_data = json.load(json_data)

In [5]:
neighborhoods_data = new_york_data['features']

In [6]:
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 
neighborhoods = pd.DataFrame(columns=column_names)
neighborhoods

Empty DataFrame
Columns: [Borough, Neighborhood, Latitude, Longitude]
Index: []

In [7]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

In [8]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

In [9]:
brooklyn_data = neighborhoods[neighborhoods['Borough'] == 'Brooklyn'].reset_index(drop=True)
brooklyn_data.head()


Borough Neighborhood   Latitude  Longitude
0  Brooklyn    Bay Ridge  40.625801 -74.030621
1  Brooklyn  Bensonhurst  40.611009 -73.995180
2  Brooklyn  Sunset Park  40.645103 -74.010316
3  Brooklyn   Greenpoint  40.730201 -73.954241
4  Brooklyn    Gravesend  40.595260 -73.973471

In [10]:
address = 'Brooklyn, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Brooklyn are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Brooklyn are 40.6501038, -73.9495823.


In [11]:
map_brooklyn = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(brooklyn_data['Latitude'], brooklyn_data['Longitude'],brooklyn_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_brooklyn)  
    
map_brooklyn

FourSquare Data

In [12]:
CLIENT_ID = 'UAU4F0LIK3WOOWV5RCM5Z2PCCYTN402X42QVDHNCOZJYEPBS' 
CLIENT_SECRET = '5DJPNJ2C2DMW0HYAMP1Y0KXDRONEQYDFQCWMAMG2LT0M4KTG' 
VERSION = '20180604'
LIMIT = 500

In [13]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [14]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [15]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [16]:
brooklyn_venues = getNearbyVenues(names=brooklyn_data['Neighborhood'],
                                   latitudes=brooklyn_data['Latitude'],
                                   longitudes=brooklyn_data['Longitude']
                                  )

Bay Ridge
Bensonhurst
Sunset Park
Greenpoint
Gravesend
Brighton Beach
Sheepshead Bay
Manhattan Terrace
Flatbush
Crown Heights
East Flatbush
Kensington
Windsor Terrace
Prospect Heights
Brownsville
Williamsburg
Bushwick
Bedford Stuyvesant
Brooklyn Heights
Cobble Hill
Carroll Gardens
Red Hook
Gowanus
Fort Greene
Park Slope
Cypress Hills
East New York
Starrett City
Canarsie
Flatlands
Mill Island
Manhattan Beach
Coney Island
Bath Beach
Borough Park
Dyker Heights
Gerritsen Beach
Marine Park
Clinton Hill
Sea Gate
Downtown
Boerum Hill
Prospect Lefferts Gardens
Ocean Hill
City Line
Bergen Beach
Midwood
Prospect Park South
Georgetown
East Williamsburg
North Side
South Side
Ocean Parkway
Fort Hamilton
Ditmas Park
Wingate
Rugby
Remsen Village
New Lots
Paerdegat Basin
Mill Basin
Fulton Ferry
Vinegar Hill
Weeksville
Broadway Junction
Dumbo
Homecrest
Highland Park
Madison
Erasmus


Check Size

In [17]:
print(brooklyn_venues.shape)
brooklyn_venues.head()

(2680, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0    Bay Ridge              40.625801              -74.030621   
1    Bay Ridge              40.625801              -74.030621   
2    Bay Ridge              40.625801              -74.030621   
3    Bay Ridge              40.625801              -74.030621   
4    Bay Ridge              40.625801              -74.030621   

                         Venue  Venue Latitude  Venue Longitude  \
0  Pilo Arts Day Spa and Salon       40.624748       -74.030591   
1                    Bagel Boy       40.627896       -74.029335   
2          Leo's Casa Calamari       40.624200       -74.030931   
3                Cocoa Grinder       40.623967       -74.030863   
4                 Pegasus Cafe       40.623168       -74.031186   

   Venue Category  
0             Spa  
1      Bagel Shop  
2     Pizza Place  
3       Juice Bar  
4  Breakfast Spot

Check how many venues were returned for each neighborhood



In [18]:
brooklyn_venues.groupby('Neighborhood').count()


Neighborhood Latitude  Neighborhood Longitude  \
Neighborhood                                                               
Bath Beach                                    48                      48   
Bay Ridge                                     79                      79   
Bedford Stuyvesant                            28                      28   
Bensonhurst                                   31                      31   
Bergen Beach                                   7                       7   
Boerum Hill                                   91                      91   
Borough Park                                  16                      16   
Brighton Beach                                45                      45   
Broadway Junction                             17                      17   
Brooklyn Heights                             100                     100   
Brownsville                                   15                      15   
Bushwick                                      71                      71   
Canarsie                                       6                       6   
Carroll Gardens                              100                     100   
City Line                                     35                      35   
Clinton Hill                                  94                      94   
Cobble Hill                                   96                      96   
Coney Island                                  15                      15   
Crown Heights                                 20                      20   
Cypress Hills                                 31                      31   
Ditmas Park                                   45                      45   
Downtown                                      90                      90   
Dumbo                                         50                      50   
Dyker Heights                                  6                       6   
East Flatbush                                 12                      12   
East New York                                 17                      17   
East Williamsburg                             71                      71   
Erasmus                                       19                      19   
Flatbush                                      21                      21   
Flatlands                                     19                      19   
Fort Greene                                   58                      58   
Fort Hamilton                                 70                      70   
Fulton Ferry                                  57                      57   
Georgetown                                    29                      29   
Gerritsen Beach                               16                      16   
Gowanus                                       63                      63   
Gravesend                                     25                      25   
Greenpoint                                    92                      92   
Highland Park                                 16                      16   
Homecrest                                     43                      43   
Kensington                                    37                      37   
Madison                                       12                      12   
Manhattan Beach                               10                      10   
Manhattan Terrace                             23                      23   
Marine Park                                   11                      11   
Midwood                                       15                      15   
Mill Basin                                    33                      33   
Mill Island                                    1                       1   
New Lots                                      22                      22   
North Side                                   100                     100   
Ocean Hill                                    24                      24   

In [19]:
print('There are {} uniques categories.'.format(len(brooklyn_venues['Venue Category'].unique())))


There are 282 uniques categories.


Analyze each area

In [20]:
# one hot encoding
brooklyn_onehot = pd.get_dummies(brooklyn_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
brooklyn_onehot['Neighborhood'] = brooklyn_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [brooklyn_onehot.columns[-1]] + list(brooklyn_onehot.columns[:-1])
brooklyn_onehot = brooklyn_onehot[fixed_columns]

brooklyn_onehot.head()

Yoga Studio  Accessories Store  American Restaurant  Animal Shelter  \
0            0                  0                    0               0   
1            0                  0                    0               0   
2            0                  0                    0               0   
3            0                  0                    0               0   
4            0                  0                    0               0   

   Antique Shop  Arepa Restaurant  Argentinian Restaurant  Art Gallery  \
0             0                 0                       0            0   
1             0                 0                       0            0   
2             0                 0                       0            0   
3             0                 0                       0            0   
4             0                 0                       0            0   

   Arts & Crafts Store  Arts & Entertainment  Asian Restaurant  \
0                    0                     0                 0   
1                    0                     0                 0   
2                    0                     0                 0   
3                    0                     0                 0   
4                    0                     0                 0   

   Athletics & Sports  Auto Dealership  Auto Garage  BBQ Joint  Bagel Shop  \
0                   0                0            0          0           0   
1                   0                0            0          0           1   
2                   0                0            0          0           0   
3                   0                0            0          0           0   
4                   0                0            0          0           0   

   Bakery  Bank  Bar  Baseball Field  Baseball Stadium  Basketball Court  \
0       0     0    0               0                 0                 0   
1       0     0    0               0                 0                 0   
2       0     0    0               0                 0                 0   
3       0     0    0               0                 0                 0   
4       0     0    0               0                 0                 0   

   Beach  Beer Bar  Beer Garden  Beer Store  Big Box Store  \
0      0         0            0           0              0   
1      0         0            0           0              0   
2      0         0            0           0              0   
3      0         0            0           0              0   
4      0         0            0           0              0   

   Bike Rental / Bike Share  Bike Shop  Bistro  Board Shop  Boat or Ferry  \
0                         0          0       0           0              0   
1                         0          0       0           0              0   
2                         0          0       0           0              0   
3                         0          0       0           0              0   
4                         0          0       0           0              0   

   Bookstore  Boutique  Boxing Gym  Breakfast Spot  Brewery  Bridge  \
0          0         0           0               0        0       0   
1          0         0           0               0        0       0   
2          0         0           0               0        0       0   
3          0         0           0               0        0       0   
4          0         0           0               1        0       0   

   Bubble Tea Shop  Buffet  Burger Joint  Burmese Restaurant  Burrito Place  \
0                0       0             0                   0              0   
1                0       0             0                   0              0   
2                0       0             0                   0              0   
3                0       0             0                   0              0   
4                0       0             0                   0              0   

   Bus Line  Bus Station  Bus Stop  Butcher  Café  C

In [21]:
brooklyn_grouped = brooklyn_onehot.groupby('Neighborhood').mean().reset_index()
brooklyn_grouped

Neighborhood  Yoga Studio  Accessories Store  \
0                  Bath Beach     0.000000           0.000000   
1                   Bay Ridge     0.012658           0.000000   
2          Bedford Stuyvesant     0.000000           0.000000   
3                 Bensonhurst     0.000000           0.000000   
4                Bergen Beach     0.000000           0.000000   
5                 Boerum Hill     0.021978           0.000000   
6                Borough Park     0.000000           0.000000   
7              Brighton Beach     0.000000           0.000000   
8           Broadway Junction     0.000000           0.000000   
9            Brooklyn Heights     0.040000           0.000000   
10                Brownsville     0.000000           0.000000   
11                   Bushwick     0.000000           0.000000   
12                   Canarsie     0.000000           0.000000   
13            Carroll Gardens     0.000000           0.000000   
14                  City Line     0.000000           0.000000   
15               Clinton Hill     0.031915           0.000000   
16                Cobble Hill     0.031250           0.000000   
17               Coney Island     0.000000           0.000000   
18              Crown Heights     0.000000           0.000000   
19              Cypress Hills     0.000000           0.000000   
20                Ditmas Park     0.022222           0.000000   
21                   Downtown     0.011111           0.000000   
22                      Dumbo     0.040000           0.000000   
23              Dyker Heights     0.000000           0.000000   
24              East Flatbush     0.000000           0.000000   
25              East New York     0.000000           0.000000   
26          East Williamsburg     0.000000           0.000000   
27                    Erasmus     0.052632           0.000000   
28                   Flatbush     0.000000           0.000000   
29                  Flatlands     0.000000           0.000000   
30                Fort Greene     0.034483           0.000000   
31              Fort Hamilton     0.000000           0.000000   
32               Fulton Ferry     0.000000           0.000000   
33                 Georgetown     0.000000           0.000000   
34            Gerritsen Beach     0.000000           0.000000   
35                    Gowanus     0.031746           0.000000   
36                  Gravesend     0.000000           0.000000   
37                 Greenpoint     0.032609           0.000000   
38              Highland Park     0.000000           0.000000   
39                  Homecrest     0.000000           0.000000   
40                 Kensington     0.000000           0.000000   
41                    Madison     0.000000           0.000000   
42            Manhattan Beach     0.000000           0.000000   
43          Manhattan Terrace     0.000000           0.000000   
44                Marine Park     0.000000           0.000000   
45                    Midwood     0.000000           0.000000   
46                 Mill Basin     0.000000           0.000000   
47                Mill Island     0.000000           0.000000   
48                   New Lots     0.000000           0.000000   
49                 North Side     0.050000           0.000000   
50                 Ocean Hill     0.000000           0.000000   
51              Ocean Parkway     0.000000           0.000000   
52            Paerdegat Basin     0.000000           0.000000   
53                 Park Slope     0.017241           0.017241   
54           Prospect Heights     0.015873           0.000000   
55  Prospect Lefferts Gardens     0.000000           0.000000   
56        Prospect Park South     0.000000           0.000000   
57                   Red Hook     0.000000           0.000000   
58             Remsen Village     0.000000           0.000000   
59                      Rugby     0.000000           0.000000   
60                   Sea Gate     0.000000          

In [22]:
num_top_venues = 10

for hood in brooklyn_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = brooklyn_grouped[brooklyn_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Bath Beach----
                       venue  freq
0                   Pharmacy  0.06
1                Gas Station  0.04
2         Chinese Restaurant  0.04
3       Fast Food Restaurant  0.04
4         Italian Restaurant  0.04
5               Dessert Shop  0.04
6            Bubble Tea Shop  0.04
7           Sushi Restaurant  0.02
8  Middle Eastern Restaurant  0.02
9      Check Cashing Service  0.02


----Bay Ridge----
                 venue  freq
0   Italian Restaurant  0.09
1                  Spa  0.06
2          Pizza Place  0.05
3                  Bar  0.04
4     Greek Restaurant  0.04
5  American Restaurant  0.04
6       Sandwich Place  0.03
7           Bagel Shop  0.03
8       Ice Cream Shop  0.03
9        Grocery Store  0.03


----Bedford Stuyvesant----
                     venue  freq
0                     Café  0.07
1                      Bar  0.07
2                 Bus Stop  0.07
3              Coffee Shop  0.07
4              Pizza Place  0.07
5              Bus Station  0.

                  venue  freq
0            Print Shop  0.08
1              Pharmacy  0.08
2           Supermarket  0.08
3         Moving Target  0.08
4     Food & Drink Shop  0.08
5  Caribbean Restaurant  0.08
6  Fast Food Restaurant  0.08
7    Chinese Restaurant  0.08
8        Hardware Store  0.08
9                  Park  0.08


----East New York----
                       venue  freq
0              Deli / Bodega  0.24
1         Spanish Restaurant  0.12
2        Fried Chicken Joint  0.06
3           Asian Restaurant  0.06
4       Fast Food Restaurant  0.06
5         Chinese Restaurant  0.06
6       Caribbean Restaurant  0.06
7              Event Service  0.06
8              Metro Station  0.06
9  Latin American Restaurant  0.06


----East Williamsburg----
                venue  freq
0                 Bar  0.08
1       Deli / Bodega  0.08
2        Cocktail Bar  0.06
3              Bakery  0.06
4         Coffee Shop  0.06
5         Music Venue  0.04
6        Concert Hall  0.04
7  Mexica



----New Lots----
                 venue  freq
0          Pizza Place  0.09
1  Fried Chicken Joint  0.09
2   Spanish Restaurant  0.09
3        Grocery Store  0.09
4       Discount Store  0.05
5       Breakfast Spot  0.05
6   Chinese Restaurant  0.05
7                 Park  0.05
8        Metro Station  0.05
9             Pharmacy  0.05


----North Side----
                           venue  freq
0                    Coffee Shop  0.10
1                    Yoga Studio  0.05
2                    Pizza Place  0.05
3                         Bakery  0.04
4            American Restaurant  0.04
5                            Bar  0.04
6                       Wine Bar  0.03
7  Vegetarian / Vegan Restaurant  0.03
8                  Jewelry Store  0.03
9                   Cocktail Bar  0.02


----Ocean Hill----
                             venue  freq
0                    Deli / Bodega  0.17
1                       Playground  0.08
2                Convenience Store  0.08
3  Southern / Soul Food Res

Top 5 most common venues above, below put in pandas data frame

In [23]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [24]:
num_top_venues = 10
indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = brooklyn_grouped['Neighborhood']

for ind in np.arange(brooklyn_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(brooklyn_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood   1st Most Common Venue  \
0                  Bath Beach                Pharmacy   
1                   Bay Ridge      Italian Restaurant   
2          Bedford Stuyvesant                     Bar   
3                 Bensonhurst                  Bakery   
4                Bergen Beach         Harbor / Marina   
5                 Boerum Hill             Coffee Shop   
6                Borough Park                    Bank   
7              Brighton Beach                   Beach   
8           Broadway Junction     Fried Chicken Joint   
9            Brooklyn Heights             Yoga Studio   
10                Brownsville              Restaurant   
11                   Bushwick                     Bar   
12                   Canarsie                     Gym   
13            Carroll Gardens      Italian Restaurant   
14                  City Line              Donut Shop   
15               Clinton Hill      Italian Restaurant   
16                Cobble Hill             Pizza Place   
17               Coney Island        Baseball Stadium   
18              Crown Heights             Pizza Place   
19              Cypress Hills     Fried Chicken Joint   
20                Ditmas Park           Deli / Bodega   
21                   Downtown            Burger Joint   
22                      Dumbo                    Park   
23              Dyker Heights            Dance Studio   
24              East Flatbush              Print Shop   
25              East New York           Deli / Bodega   
26          East Williamsburg                     Bar   
27                    Erasmus    Caribbean Restaurant   
28                   Flatbush               Juice Bar   
29                  Flatlands                Pharmacy   
30                Fort Greene      Italian Restaurant   
31              Fort Hamilton      Italian Restaurant   
32               Fulton Ferry                    Park   
33                 Georgetown                    Bank   
34            Gerritsen Beach                     Bar   
35                    Gowanus  Furniture / Home Store   
36                  Gravesend             Pizza Place   
37                 Greenpoint             Pizza Place   
38              Highland Park      Spanish Restaurant   
39                  Homecrest                    Bank   
40                 Kensington         Thai Restaurant   
41                    Madison           Deli / Bodega   
42            Manhattan Beach                    Café   
43          Manhattan Terrace             Pizza Place   
44                Marine Park           Deli / Bodega   
45                    Midwood             Pizza Place   
46                 Mill Basin             Pizza Place   
47                Mill Island                    Pool   
48                   New Lots             Pizza Place   
49                 North Side             Coffee Shop   
50                 Ocean Hill           Deli / Bodega   
51              Ocean Parkway              Restaurant   
52            Paerdegat Basin                     Gym   
53                 Park Slope            Burger Joint   
54           Prospect Heights                     Bar   
55  Prospect Lefferts Gardens                  Bakery   
56        Prospect Park South    Caribbean Restaurant   
57                   Red Hook      Seafood Restaurant   
58             Remsen Village    Caribbean Restaurant   
59                      Rugby    Caribbean Restaurant   
60                   Sea Gate             Sports Club   
61             Sheepshead Bay            Dessert Shop   
62                 South Side                     Bar   
63              Starrett City           Moving Target   
64                Sunset Park                    Bank   
65               Vinegar Hill              Food Truck   
66                 Weeksville           Deli / Bodega   
67               Williamsburg             Coffee Shop   
68            Windsor Terrace                   Diner   
69                    Winga

Cluster Neighborhoods

In [25]:
# set number of clusters
kclusters = 4

brooklyn_grouped_clustering = brooklyn_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(brooklyn_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 3, 0, 0, 0])

In [26]:
brooklyn_merged = brooklyn_data

# add clustering labels
brooklyn_merged['Cluster Labels'] = kmeans.labels_

brooklyn_merged = brooklyn_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

brooklyn_merged.head(15) # check the last columns!

Borough       Neighborhood   Latitude  Longitude  Cluster Labels  \
0   Brooklyn          Bay Ridge  40.625801 -74.030621               0   
1   Brooklyn        Bensonhurst  40.611009 -73.995180               0   
2   Brooklyn        Sunset Park  40.645103 -74.010316               0   
3   Brooklyn         Greenpoint  40.730201 -73.954241               0   
4   Brooklyn          Gravesend  40.595260 -73.973471               0   
5   Brooklyn     Brighton Beach  40.576825 -73.965094               0   
6   Brooklyn     Sheepshead Bay  40.586890 -73.943186               3   
7   Brooklyn  Manhattan Terrace  40.614433 -73.957438               0   
8   Brooklyn           Flatbush  40.636326 -73.958401               0   
9   Brooklyn      Crown Heights  40.670829 -73.943291               0   
10  Brooklyn      East Flatbush  40.641718 -73.936103               2   
11  Brooklyn         Kensington  40.642382 -73.980421               0   
12  Brooklyn    Windsor Terrace  40.656946 -73.980073               3   
13  Brooklyn   Prospect Heights  40.676822 -73.964859               0   
14  Brooklyn        Brownsville  40.663950 -73.910235               0   

   1st Most Common Venue 2nd Most Common Venue      3rd Most Common Venue  \
0     Italian Restaurant                   Spa                Pizza Place   
1                 Bakery      Sushi Restaurant              Grocery Store   
2                   Bank                Bakery  Latin American Restaurant   
3            Pizza Place                   Bar                Coffee Shop   
4            Pizza Place           Bus Station              Deli / Bodega   
5                  Beach    Russian Restaurant                 Restaurant   
6           Dessert Shop    Turkish Restaurant             Sandwich Place   
7            Pizza Place            Donut Shop             Ice Cream Shop   
8              Juice Bar  Caribbean Restaurant                   Pharmacy   
9            Pizza Place                Museum                       Café   
10            Print Shop              Pharmacy       Caribbean Restaurant   
11       Thai Restaurant         Grocery Store             Sandwich Place   
12                 Diner         Grocery Store                       Café   
13                   Bar    Mexican Restaurant                     Bakery   
14            Restaurant    Chinese Restaurant                       Park   

          4th Most Common Venue 5th Most Common Venue  6th Most Common Venue  \
0           American Restaurant                   Bar       Greek Restaurant   
1            Italian Restaurant        Ice Cream Shop             Donut Shop   
2            Mexican Restaurant           Pizza Place         Ice Cream Shop   
3                  Cocktail Bar           Yoga Studio       Sushi Restaurant   
4                        Lounge                Bakery     Italian Restaurant   
5   Eastern European Restaurant          Gourmet Shop       Sushi Restaurant   
6                    Bagel Shop                 Diner               Creperie   
7                     Jazz Club          Liquor Store     Chinese Restaurant   
8            Mexican Restaurant           Coffee Shop             Donut Shop   
9                 Grocery Store            Playground            Supermarket   
10            Food & Drink Shop    Chinese Restaurant                   Park   
11                  Pizza Place        Ice Cream Shop                    Spa   
12                Deli / Bodega                  Park             Bagel Shop   
13                 Cocktail Bar             Wine Shop        Thai Restaurant   
14            Convenience Store   Fried Chicken Joint  Performing Arts Venue   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0             Bagel Shop        Sandwich Place            Playground   
1     Chinese Restaurant          Noodle House    Russian Restaurant   
2                    Gym   Fried Chicken Joint     Mobile Phone Shop   
3                   Café          

In [27]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(brooklyn_merged['Latitude'], brooklyn_merged['Longitude'], brooklyn_merged['Neighborhood'], brooklyn_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Look at all the clusters

Cluster 1

In [28]:
brooklyn_merged.loc[brooklyn_merged['Cluster Labels'] == 0, brooklyn_merged.columns[[1] + list(range(5, brooklyn_merged.shape[1]))]]

Neighborhood   1st Most Common Venue  \
0                   Bay Ridge      Italian Restaurant   
1                 Bensonhurst                  Bakery   
2                 Sunset Park                    Bank   
3                  Greenpoint             Pizza Place   
4                   Gravesend             Pizza Place   
5              Brighton Beach                   Beach   
7           Manhattan Terrace             Pizza Place   
8                    Flatbush               Juice Bar   
9               Crown Heights             Pizza Place   
11                 Kensington         Thai Restaurant   
13           Prospect Heights                     Bar   
14                Brownsville              Restaurant   
15               Williamsburg             Coffee Shop   
16                   Bushwick                     Bar   
18           Brooklyn Heights             Yoga Studio   
20            Carroll Gardens      Italian Restaurant   
21                   Red Hook      Seafood Restaurant   
22                    Gowanus  Furniture / Home Store   
23                Fort Greene      Italian Restaurant   
26              East New York           Deli / Bodega   
30                Mill Island                    Pool   
31            Manhattan Beach                    Café   
32               Coney Island        Baseball Stadium   
33                 Bath Beach                Pharmacy   
34               Borough Park                    Bank   
35              Dyker Heights            Dance Studio   
36            Gerritsen Beach                     Bar   
37                Marine Park           Deli / Bodega   
39                   Sea Gate             Sports Club   
40                   Downtown            Burger Joint   
41                Boerum Hill             Coffee Shop   
42  Prospect Lefferts Gardens                  Bakery   
43                 Ocean Hill           Deli / Bodega   
45               Bergen Beach         Harbor / Marina   
46                    Midwood             Pizza Place   
49          East Williamsburg                     Bar   
51                 South Side                     Bar   
52              Ocean Parkway              Restaurant   
53              Fort Hamilton      Italian Restaurant   
54                Ditmas Park           Deli / Bodega   
55                    Wingate     Fried Chicken Joint   
57             Remsen Village    Caribbean Restaurant   
60                 Mill Basin             Pizza Place   
61               Fulton Ferry                    Park   
62               Vinegar Hill              Food Truck   
64          Broadway Junction     Fried Chicken Joint   
65                      Dumbo                    Park   
67              Highland Park      Spanish Restaurant   
68                    Madison           Deli / Bodega   

              2nd Most Common Venue         3rd Most Common Venue  \
0                               Spa                   Pizza Place   
1                  Sushi Restaurant                 Grocery Store   
2                            Bakery     Latin American Restaurant   
3                               Bar                   Coffee Shop   
4                       Bus Station                 Deli / Bodega   
5                Russian Restaurant                    Restaurant   
7                        Donut Shop                Ice Cream Shop   
8              Caribbean Restaurant                      Pharmacy   
9                            Museum                          Café   
11                    Grocery Store                Sandwich Place   
13               Mexican Restaurant                        Bakery   
14               Chinese Restaurant                          Park   
15                       Bagel Shop                           Bar   
16               Mexican Restaurant                   Coffee Shop   
18                    Deli / Bodega                          Park   
20                      Coffee Shop                   Pizza Pl

Cluster 2

In [29]:
brooklyn_merged.loc[brooklyn_merged['Cluster Labels'] == 1, brooklyn_merged.columns[[1] + list(range(5, brooklyn_merged.shape[1]))]]


Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
47  Prospect Park South  Caribbean Restaurant  Fast Food Restaurant   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
47         Grocery Store           Pizza Place     Mobile Phone Shop   

        6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
47  Latin American Restaurant   Fried Chicken Joint    Mexican Restaurant   

   9th Most Common Venue 10th Most Common Venue  
47            Kids Store               Pharmacy

Cluster 3

In [30]:
brooklyn_merged.loc[brooklyn_merged['Cluster Labels'] == 2, brooklyn_merged.columns[[1] + list(range(5, brooklyn_merged.shape[1]))]]

Neighborhood 1st Most Common Venue      2nd Most Common Venue  \
10  East Flatbush            Print Shop                   Pharmacy   
19    Cobble Hill           Pizza Place                        Bar   
25  Cypress Hills   Fried Chicken Joint  Latin American Restaurant   
38   Clinton Hill    Italian Restaurant                Pizza Place   
44      City Line            Donut Shop          Mobile Phone Shop   
48     Georgetown                  Bank                   Pharmacy   
50     North Side           Coffee Shop                Yoga Studio   
66      Homecrest                  Bank              Grocery Store   
69        Erasmus  Caribbean Restaurant                Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
10  Caribbean Restaurant     Food & Drink Shop    Chinese Restaurant   
19            Playground           Coffee Shop    Italian Restaurant   
25           Pizza Place        Ice Cream Shop            Donut Shop   
38       Thai Restaurant             Wine Shop    Mexican Restaurant   
44                Bakery            Shoe Store         Grocery Store   
48        Breakfast Spot            Donut Shop           Coffee Shop   
50           Pizza Place   American Restaurant                Bakery   
66            Donut Shop    Mexican Restaurant    Chinese Restaurant   
69  Gym / Fitness Center            Playground           Pizza Place   

   6th Most Common Venue  7th Most Common Venue  \
10                  Park   Fast Food Restaurant   
19         Deli / Bodega           Cocktail Bar   
25    Chinese Restaurant   Fast Food Restaurant   
38     Indian Restaurant     Chinese Restaurant   
44    Spanish Restaurant   Fast Food Restaurant   
48    Miscellaneous Shop           Burger Joint   
50                   Bar               Wine Bar   
66      Sushi Restaurant  Vietnamese Restaurant   
69     Convenience Store               Pharmacy   

            8th Most Common Venue      9th Most Common Venue  \
10                 Hardware Store                Supermarket   
19                    Yoga Studio  Middle Eastern Restaurant   
25                  Metro Station         Spanish Restaurant   
38                  Grocery Store        Japanese Restaurant   
44            Fried Chicken Joint         Mexican Restaurant   
48             Mexican Restaurant             Shipping Store   
50  Vegetarian / Vegan Restaurant              Jewelry Store   
66                 Sandwich Place                Pizza Place   
69                     Donut Shop                   Bus Line   

       10th Most Common Venue  
10              Moving Target  
19                     Bakery  
25         Mexican Restaurant  
38                 Restaurant  
44  South American Restaurant  
48         Frozen Yogurt Shop  
50         Mexican Restaurant  
66         Russian Restaurant  
69                Music Venue

Cluster 4

In [31]:
brooklyn_merged.loc[brooklyn_merged['Cluster Labels'] == 3, brooklyn_merged.columns[[1] + list(range(5, brooklyn_merged.shape[1]))]]


Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
6       Sheepshead Bay          Dessert Shop    Turkish Restaurant   
12     Windsor Terrace                 Diner         Grocery Store   
17  Bedford Stuyvesant                   Bar              Bus Stop   
24          Park Slope          Burger Joint           Coffee Shop   
27       Starrett City         Moving Target           Pizza Place   
28            Canarsie                   Gym              Bus Line   
29           Flatlands              Pharmacy  Caribbean Restaurant   
56               Rugby  Caribbean Restaurant         Grocery Store   
58            New Lots           Pizza Place    Spanish Restaurant   
59     Paerdegat Basin                   Gym      Asian Restaurant   
63          Weeksville         Deli / Bodega        Discount Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6         Sandwich Place            Bagel Shop                 Diner   
12                  Café         Deli / Bodega                  Park   
17           Pizza Place                  Café           Coffee Shop   
24            Bagel Shop     Korean Restaurant                   Pub   
27              Bus Stop            Donut Shop              Pharmacy   
28    Chinese Restaurant  Caribbean Restaurant      Asian Restaurant   
29   Fried Chicken Joint  Fast Food Restaurant                Lounge   
56                  Bank    Seafood Restaurant        Sandwich Place   
58         Grocery Store   Fried Chicken Joint    Chinese Restaurant   
59         Moving Target                  Food       Harbor / Marina   
63         Grocery Store            Donut Shop          Cocktail Bar   

   6th Most Common Venue 7th Most Common Venue          8th Most Common Venue  \
6               Creperie          Outlet Store                     Restaurant   
12            Bagel Shop    Chinese Restaurant                     Beer Store   
17             Juice Bar           Bus Station                       Boutique   
24           Pizza Place             Pet Store                Thai Restaurant   
27   American Restaurant  Caribbean Restaurant                           Food   
28         Grocery Store           Fish Market              Fish & Chips Shop   
29                   Bar     Electronics Store  Paper / Office Supplies Store   
56           Pizza Place           Wings Joint             Salon / Barbershop   
58        Breakfast Spot    Salon / Barbershop      Latin American Restaurant   
59              Bus Line    Falafel Restaurant                           Farm   
63                  Park    Chinese Restaurant                           Café   

      9th Most Common Venue 10th Most Common Venue  
6        Russian Restaurant                   Café  
12              Coffee Shop              Bookstore  
17  New American Restaurant              Gift Shop  
24                   Bakery              Bookstore  
27        Food & Drink Shop            Flower Shop  
28      Filipino Restaurant                  Field  
29                     Park              Nightclub  
56                    Diner      Mobile Phone Shop  
58   Furniture / Home Store                   Bank  
59           Farmers Market   Fast Food Restaurant  
63                     Food            Gas Station

Park Slope in cluster 4 and Williamsburg in cluster 1

In [32]:
cluster4 = brooklyn_merged.loc[brooklyn_merged['Cluster Labels'] == 3][['Neighborhood']]
brooklyn_cluster4 = cluster4.loc[cluster4['Neighborhood'] != 'Park Slope'].reset_index(drop=True)
print("The similar neighborhoods in Brooklyn are:")
brooklyn_cluster4

The similar neighborhoods in Brooklyn are:


Neighborhood
0      Sheepshead Bay
1     Windsor Terrace
2  Bedford Stuyvesant
3       Starrett City
4            Canarsie
5           Flatlands
6               Rugby
7            New Lots
8     Paerdegat Basin
9          Weeksville

now lets get the area with the lowest rent

In [33]:
rent_bk

Neighborhood  Average Rent
0                             Flatlands          1492
1                           Marine Park          1659
2                               Madison          1659
3                       Gerritsen Beach          1659
4                             Homecrest          1717
5                           Bensonhurst          1795
6                            Bath Beach          1795
7                               Midwood          1825
8                         Fiske Terrace          1825
9                            Kensington          1839
10                            Gravesend          1840
11                        Crown Heights          1853
12                         Borough Park          1898
13                          Ditmas Park          1944
14                  Prospect Park South          1997
15          Prospect - Lefferts Gardens          2009
16                             Flatbush          2009
17                              Wingate          2114
18                       Sheepshead Bay          2131
19                        East Flatbush          2178
20                             Canarsie          2178
21                      Manhattan Beach          2214
22                       Brighton Beach          2214
23                      Windsor Terrace          2361
24                         Clinton Hill          2671
25                   Bedford-Stuyvesant          2784
26                        Prospect Park          2876
27                             Red Hook          2916
28                      Carroll Gardens          2916
29                             Bushwick          3065
30                   Brooklyn Navy Yard          3136
31                              Gowanus          3247
32                         Williamsburg          3247
33                     Prospect Heights          3373
34  Columbia Street Waterfront District          3401
35                          Cobble Hill          3401
36                          Fort Greene          3403
37                         Vinegar Hill          3456
38                                DUMBO          3456
39                    Downtown Brooklyn          3456
40                     Brooklyn Heights          3456
41                          Boerum Hill          3462
42                           Park Slope          3513
43                    Greenwood Heights          3538

In [34]:
pd.concat([brooklyn_cluster4, rent_bk], axis=1, join='inner')
cluster4_rent = pd.merge(brooklyn_cluster4, rent_bk,  how='left', left_on=['Neighborhood'], right_on = ['Neighborhood']).dropna()
cluster4_rent_asc = cluster4_rent.sort_values('Average Rent', axis=0, ascending = 'True').reset_index(drop=True)
cluster4_rent_asc

Neighborhood  Average Rent
0        Flatlands        1492.0
1   Sheepshead Bay        2131.0
2         Canarsie        2178.0
3  Windsor Terrace        2361.0

Flatlands is the 1st contender, now for cluster 1

In [35]:
cluster1 = brooklyn_merged.loc[brooklyn_merged['Cluster Labels'] == 0][['Neighborhood']]
brooklyn_cluster1 = cluster1.loc[cluster1['Neighborhood'] != 'Williamsburg'].reset_index(drop=True)
print("The similar neighborhoods in Brooklyn are:")
brooklyn_cluster1

The similar neighborhoods in Brooklyn are:


Neighborhood
0                   Bay Ridge
1                 Bensonhurst
2                 Sunset Park
3                  Greenpoint
4                   Gravesend
5              Brighton Beach
6           Manhattan Terrace
7                    Flatbush
8               Crown Heights
9                  Kensington
10           Prospect Heights
11                Brownsville
12                   Bushwick
13           Brooklyn Heights
14            Carroll Gardens
15                   Red Hook
16                    Gowanus
17                Fort Greene
18              East New York
19                Mill Island
20            Manhattan Beach
21               Coney Island
22                 Bath Beach
23               Borough Park
24              Dyker Heights
25            Gerritsen Beach
26                Marine Park
27                   Sea Gate
28                   Downtown
29                Boerum Hill
30  Prospect Lefferts Gardens
31                 Ocean Hill
32               Bergen Beach
33                    Midwood
34          East Williamsburg
35                 South Side
36              Ocean Parkway
37              Fort Hamilton
38                Ditmas Park
39                    Wingate
40             Remsen Village
41                 Mill Basin
42               Fulton Ferry
43               Vinegar Hill
44          Broadway Junction
45                      Dumbo
46              Highland Park
47                    Madison

In [36]:
pd.concat([brooklyn_cluster1, rent_bk], axis=1, join='inner')
cluster1_rent = pd.merge(brooklyn_cluster1, rent_bk,  how='left', left_on=['Neighborhood'], right_on = ['Neighborhood']).dropna()
cluster1_rent_asc = cluster1_rent.sort_values('Average Rent', axis=0, ascending = 'True').reset_index(drop=True)
cluster1_rent_asc

Neighborhood  Average Rent
0            Madison        1659.0
1        Marine Park        1659.0
2    Gerritsen Beach        1659.0
3         Bath Beach        1795.0
4        Bensonhurst        1795.0
5            Midwood        1825.0
6         Kensington        1839.0
7          Gravesend        1840.0
8      Crown Heights        1853.0
9       Borough Park        1898.0
10       Ditmas Park        1944.0
11          Flatbush        2009.0
12           Wingate        2114.0
13   Manhattan Beach        2214.0
14    Brighton Beach        2214.0
15          Red Hook        2916.0
16   Carroll Gardens        2916.0
17          Bushwick        3065.0
18           Gowanus        3247.0
19  Prospect Heights        3373.0
20       Fort Greene        3403.0
21  Brooklyn Heights        3456.0
22      Vinegar Hill        3456.0
23       Boerum Hill        3462.0

We can see that Madison has the lowest rent in cluster 1. The 2nd contender. 

In [37]:
flatlands = brooklyn_data.loc[29,'Neighborhood']

In [38]:
madison = brooklyn_data.loc[68,'Neighborhood']

Lets display them on a map

In [39]:
#hotplace_df = brooklyn_data.drop('Neighborhood' != 'Madison' or 'Neighborhood' != 'Flatlands', axis = True)
trendy_df = brooklyn_data[brooklyn_data.Neighborhood == 'Madison']
trendy_df = trendy_df.append(brooklyn_data[brooklyn_data.Neighborhood == 'Flatlands'])


trendy_df.head()

Borough Neighborhood   Latitude  Longitude  Cluster Labels
68  Brooklyn      Madison  40.609378 -73.948415               0
29  Brooklyn    Flatlands  40.630446 -73.929113               3

In [40]:
new_map = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(trendy_df['Latitude'], trendy_df['Longitude'],trendy_df['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(new_map)  
    
new_map

**Look back at previous data:**

----Flatlands----
                           venue  freq
0                       Pharmacy  0.16
1           Fast Food Restaurant  0.11
2           Caribbean Restaurant  0.11
3            Fried Chicken Joint  0.11
4  Paper / Office Supplies Store  0.05
5                           Park  0.05
6                      Nightclub  0.05
7                            Bar  0.05
8                         Lounge  0.05
9                   Intersection  0.05




----Madison----
                venue  freq
0       Deli / Bodega  0.17
1          Bagel Shop  0.17
2          Hobby Shop  0.08
3      Pilates Studio  0.08
4         Pizza Place  0.08
5        Dessert Shop  0.08
6         Candy Store  0.08
7         Bus Station  0.08
8                 Spa  0.08
9  Italian Restaurant  0.08




You could see that Flatlands has more exciting venues such as Nightclub, Bar, and Lounge, while Madison has more choice based on food venues. I would make sure to present this data to the investor and while very subjective Flatlands seems like the more exciting Neighborhood.

# Summary: 


This project found that the closest related neighborhood to Park Slope with the lowest possible rent was Flatlands (rent: 1492.0 per month) 
and the closest related neighborhood to Williamsburg with lowest possible rent was Madison (rent:1659.0 per month) 
After mapping the two areas, we see that they are in similar locations, but Flatlands has the lower rent by far. 
This Neighborhood could experience the success that Park Slope experienced in the future, as Hipsters and Yuppies get priced out of Park Slope, Flatlands could be their next destination. Thus invesment in Flatlands could be suggested. 
